### PR TITLE
Core/Transmog: Add TransmogSets

### LINKED ISSUE #20133: Transmog Sets
**Description:**

- Achievements for completing transmog sets is missing
- SPELL_EFFECT_LEARN_TRANSMOG_SET is NYI

**Current behaviour:**

See above

**Expected behaviour:**

SPELL_EFFECT_LEARN_TRANSMOG_SET should learn you the appearances for the set
When you complete a set, the achievement should be granted to you

**Steps to reproduce the problem:**

1. Create a plate wearing character
2. cast spell 247630 (Ensemble: Timewarden's Plate) given by chromie quests
3. Nothing happens

**Branch:**

master

**TC rev. hash/commit:** 
https://github.com/TrinityCore/TrinityCore/commit/cb4dd92629149d94e5ded6c2e8b073173bbe042a
**TDB version:**  7x
**Operating system:** Win 10

**Notes:**
I got something locally, but I don't like how I made it, not feeling good sharing it now (maybe tomorrow)

### PR DESCRIPTION
**Pull request by Golrag**
**Changes proposed:**

-  Add Achievement for transmog set completion
-  SPELL_EFFECT_LEARN_TRANSMOG_SET

**Target branch(es):**
- master

**Issues addressed:** Closes #20133

**Tests performed:**
- Tested in game
- Compiles

**Known issues and TODO list:** (add/remove lines as needed)

### CODE REVIEW

**Comment by Shoxxo** [comment]
Tested ...

[Working](https://image.prntscr.com/image/33FDOv_KSZSeDM17qVPiRA.png)

**Comment by Golrag** [comment]
Oh I am talking about the tab 'Sets' in your appearances collection not about the transmogging to a set.
I could take a look at that tho and see if I it fits this PR

**Comment by Shoxxo** [comment]
Yes i know and as you can see, it is avaible or i am wrong?

**Comment by Golrag** [comment]
Yea it's available :)

File: src/server/game/Entities/Player/CollectionMgr.cpp [hunk 1]
```diff
@@ -603,6 +603,58 @@ void CollectionMgr::AddItemAppearance(uint32 itemId, uint32 appearanceModId /*=
...
+    {
+        ItemModifiedAppearanceEntry const* itemModifiedAppearance = sItemModifiedAppearanceStore.LookupEntry(item->ItemModifiedAppearanceID);
+        if (!itemModifiedAppearance)
+            continue;
+
+        AddItemAppearance(itemModifiedAppearance);
+    }
+}
+
+bool CollectionMgr::IsSetCompleted(uint32 transmogSetId) const
```
Comment (Shauren) [inline src/server/game/Entities/Player/CollectionMgr.cpp]: https://gist.github.com/Shauren/6929981f075b8a40b6c822a209fd518a
Improved version with no temporary containers (this is what the client does too btw with ItemTransmogrificationSlots)

File: src/server/game/DataStores/DB2Stores.cpp [hunk 2]
```diff
@@ -1958,6 +1978,24 @@ bool DB2Manager::IsToyItem(uint32 toy) const
     return _toys.count(toy) > 0;
 }
 
+std::vector<TransmogSetEntry const*> const* DB2Manager::GetTransmogSetsForItemModifiedAppearance(uint32 itemModID) const
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.cpp]: Please use full name for argument (itemModifiedAppearanceId)

**Review by Shauren — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Player/CollectionMgr.cpp [hunk 3]
```diff
@@ -603,6 +603,50 @@ void CollectionMgr::AddItemAppearance(uint32 itemId, uint32 appearanceModId /*=
...
+    std::array<int8, EQUIPMENT_SLOT_END> knownPieces;
+    knownPieces.fill(-1);
+    for (TransmogSetItemEntry const* transmogSetItem : *transmogSetItems)
+    {
+        ItemModifiedAppearanceEntry const* itemModifiedAppearance = sItemModifiedAppearanceStore.LookupEntry(transmogSetItem->ItemModifiedAppearanceID);
+        if (!itemModifiedAppearance)
+            continue;
+
+        ItemEntry const* item = sItemStore.LookupEntry(itemModifiedAppearance->ItemID);
+        int32 transmogSlot = ItemTransmogrificationSlots[item->InventoryType];
```
Comment (Golrag) [inline src/server/game/Entities/Player/CollectionMgr.cpp]: @Shauren shouldn't we check if item is null or is it guaranteed to exist ?

**Comment by Golrag** [comment]
I could take a look at CRITERIA_TYPE_APPEARANCE_UNLOCKED_BY_SLOT in this PR btw maybe it fits :p

File: src/server/game/Entities/Player/CollectionMgr.cpp (hunk 3 above)
Comment (Shauren) [inline src/server/game/Entities/Player/CollectionMgr.cpp]: You should check it - i didn't do complete checking there

File: src/server/game/Entities/Item/Item.h [hunk 4]
```diff
@@ -124,6 +124,39 @@ struct ItemDynamicFieldGems
 };
 #pragma pack(pop)
 
+int32 const ItemTransmogrificationSlots[MAX_INVTYPE] =
```
Comment (Shauren) [inline src/server/game/Entities/Item/Item.h]: This change is bad, all you had to do here was write a declaration - `extern int32 const ItemTransmogrificationSlots[MAX_INVTYPE]`

**Comment by Golrag** [comment]
Added CRITERIA_TYPE_APPEARANCE_UNLOCKED_BY_SLOT but somehow progress is not shown just like with issue #18448

**Pull request closed without merging**